brain/synapses.cpp

```diff
@@ -0,0 +1,592 @@
+                postCenterPositionZ.get()[i] = pos[j][12];
+                ++i;
+            }
+        }
+
+        if( !haveSize )
+        {
+            if( !_afferent )
+                preGID.swap( postGID );
+            _size = i;
+        }
+    }
+
+    void _allocateAttributes( const size_t size, const bool allocateIndex ) const
+    {
+        if( allocateIndex )
+            allocate( index, size );
+
+        allocate( preGID, size );
+        allocate( preSectionID, size );
+        allocate( preSegmentID, size );
+        allocate( preDistance, size );
+
+        allocate( postGID, size );
+        allocate( postSectionID, size );
+        allocate( postSegmentID, size );
+        allocate( postDistance, size );
+
+        allocate( delay, size );
+        allocate( conductance, size );
+        allocate( utilization, size );
+        allocate( depression, size );
+        allocate( facilitation, size );
+        allocate( decay, size );
+        allocate( efficacy, size );
+    }
+
+    void _allocatePositions( const size_t size ) const
+    {
+        allocate( preGID, size );
+        allocate( preSurfacePositionX, size );
+        allocate( preSurfacePositionY, size );
+        allocate( preSurfacePositionZ, size );
+        allocate( preCenterPositionX, size );
+        allocate( preCenterPositionY, size );
+        allocate( preCenterPositionZ, size );
+
+        allocate( postGID, size );
+        allocate( postSurfacePositionX, size );
+        allocate( postSurfacePositionY, size );
+        allocate( postSurfacePositionZ, size );
+        allocate( postCenterPositionX, size );
+        allocate( postCenterPositionY, size );
+        allocate( postCenterPositionZ, size );
+    }
+
+    void _ensureAttributes() const
+    {
+        if( _hasAttributes( ))
+            return;
+
+        lunchbox::ScopedWrite mutex( _lock );
+        _loadAttributes( _afferent ? _postGIDs : _preGIDs,
+                         _afferent ? _preGIDs : _postGIDs );
+    }
+
+    void _ensurePositions() const
+    {
+        if( _hasPositions( ))
+            return;
+
+        lunchbox::ScopedWrite mutex( _lock );
+        _loadPositions( _afferent ? _postGIDs : _preGIDs,
+                        _afferent ? _preGIDs : _postGIDs );
+    }
+
+    bool _hasAttributes() const
+    {
+        lunchbox::ScopedRead mutex( _lock );
+        return !!efficacy;
+    }
+
+    bool _hasPositions() const
+    {
+        lunchbox::ScopedRead mutex( _lock );
+        return !!preSurfacePositionX;
+    }
+
+    const Circuit& _circuit;
+    const GIDSet _preGIDs;
+    const GIDSet _postGIDs;
+    const bool _afferent;
+
+    template<typename T>
+    struct FreeDeleter
+    {
+        void operator()( T* ptr ) { free( ptr ); }
+    };
+
+    mutable size_t _size;
+
+    mutable std::unique_ptr< size_t, FreeDeleter<size_t> > index;
+
+    mutable std::unique_ptr< uint32_t, FreeDeleter<uint32_t> > preGID;
```

Comment:
typedef std::unique_ptr< uint32_t, FreeDeleter<uint32_t> uint32_ts; ? (and the same for the rest)